### ISSUE TITLE
k/k is the canonical source of all build artifacts.

### ISSUE THREAD

**Issue by timothysc** [kubernetes/kubernetes#71677]
xref: https://github.com/kubernetes/kubernetes/issues/71530

The release-repo contains aspects of the release that we know it should not own.  For example, it owns the publishing of .deb and .rpm artifacts and they are not properly maintained and versioned.  This has been the state of affairs for far too long. 

This issue exists as a tracking/umbrella issue to factor out the build artifacts such that k/k contains all the necessary tools+scripts+etc in order to build these artifacts.  Signing and publishing should likely still be a separate step.  

The following are aspirational requirements that we will need to support this effort
- [x] Investigate simplified build tooling that supports both rpm+.deb artifact building from a single conf-file (https://github.com/goreleaser/nfpm) 
- [ ] Create a builder container w/Dockerfile that can be used as part of the build.
- [ ] decouple bazel to strictly function as a variable override for a smaller set of kubernetes specific steps.  (the current state is conflating and pathological) 
    - For Example:
        - 1. conf step - nfpm.yaml.in -> bazel -> nfpm.yaml(out) 
        - 2. build step - nfpm.yaml -> (Builder Container) -> .debs & rpms. (could be linked to bazel as a separate step) 
- [ ] Update release repo to remove artifact building but to consume artifacts from the build process 
- [ ] BONUS: - Create a meta-installer container which host the rpm/deb repo to allow for a local install via local script on entry:  `e.g. docker run kubeadmer`

There are several more possible steps we could do, but this is an MVP to decouple.  

/kind feature

**Comment by timothysc** [comment]
![image](https://user-images.githubusercontent.com/169553/49409383-0cc73580-f726-11e8-923a-3411f224b7fd.png)

**Comment by mauilion** [comment]
linking https://github.com/kubernetes/kubernetes/pull/71686

**Comment by ixdy** [comment]
Is there actually any documentation on nfpm? the README is extremely sparse and doesn't really indicate how to use it or what it does or doesn't support.

**Comment by ixdy** [comment]
I installed nfpm and ran `nfpm init` and the generated `nfpm.yaml` looks nothing like that:
```yaml
# nfpm example config file
name: "foo"
arch: "amd64"
platform: "linux"
version: "v${MY_APP_VERSION}"
section: "default"
priority: "extra"
replaces:
- foobar
provides:
- bar
depends:
- foo
- bar
# recommends on rpm packages requires rpmbuild >= 4.13
recommends:
- whatever
# suggests on rpm packages requires rpmbuild >= 4.13
suggests:
- something-else
conflicts:
- not-foo
- not-bar
maintainer: "John Doe <john@example.com>"
description: |
  FooBar is the great foo and bar software.
    And this can be in multiple lines!
vendor: "FooBarCorp"
homepage: "http://eaxmple.com"
license: "MIT"
bindir: "/usr/local/bin"
files:
  ./foo: "/usr/local/bin/foo"
  ./bar: "/usr/local/bin/bar"
config_files:
  ./foobar.conf: "/etc/foobar.conf"
overrides:
  rpm:
    scripts:
      preinstall: ./scripts/preinstall.sh
      postremove: ./scripts/postremove.sh
  deb:
    scripts:
      postinstall: ./scripts/postinstall.sh
      preremove: ./scripts/preremove.sh
```

(some parts are kinda similar, but things are different enough to be worrying)

**Comment by ixdy** [comment]
I'm not sure if this is the right issue, but there's also the meta-question of "why are we even building debs and RPMs?"

**Comment by mauilion** [comment]
In rereading the thing I linked is the doc for goreleasers nfpm integration. Not for nfpm directly. 

I'm not committed to this tooling there are other perhaps better options out there. I do like that this uses a single config and syntax to build tar rpm or debs.

**Comment by luxas** [comment]
> I'm not sure if this is the right issue, but there's also the meta-question of "why are we even building debs and RPMs?"

In one word, to distribute the kubelet and its deps. The only 100% working solution of running the kubelet is on-host using e.g. systemd, so that's why we ship that, along with our CLI tools kubectl and kubeadm.

If we don't want to use fpm or nfpm, we can always consider using bazel to build the packages, in an "extracted" kind of way.

**Comment by neolit123** [comment]
@timothysc here is an experimental container that builds kubeadm and the kubelet:
https://github.com/neolit123/kubeadm-builder

we spoke about that on Tuesday, or was it Wednesday.

**Comment by dims** [comment]
we now have `go get github.com/kubernetes/kubernetes/cmd/kubeadm` and `go get github.com/kubernetes/kubernetes/cmd/kubelet` working! :)

**Comment by fabriziopandini** [comment]
Another thing to be improved is automatic generation of .debs/.rpm for alpha/beta/RC releases because currently you can have docker images for testing those releases, but no debs/.rpm are available

**Comment by timothysc** [comment]
cc @Klaven

**Comment by mbohlool** [comment]
If the RPMs and DEBs are published the same way other official binaries are published (like [the ones here](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.12.md#downloads-for-v1125), basically another download link set for them) I think it would be very simple to then publish them to the official RPM/DEB repos. First step (and only step we may need) is an automated tool running inside Google to look for those downloadable RPMS/DEBS and publish them to the right repo. Second step is to move those repos to a CNCF repo.

**Comment by timothysc** [comment]
@Klaven is currently work working on this, and those who are interested, please attend the kubeadm call https://github.com/kubernetes/community/tree/master/sig-cluster-lifecycle for more details.  As we progress, PRs will reference this issue.

**Comment by bart0sh** [comment]
FWIW here is the script that builds kubeadm from the current kubernetes tree and builds kubeadm deb package inside the container using Debian native tools. The build takes around 40 seconds on my old Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz desktop machine.
As you see most of the code is generating debian/ content. I didn't check, but it should work for Ubuntu too.

Building rpm would be even simpler, considering that all we need is to generate a spec.

```
#!/bin/sh

VERSION=1.14.0~alpha.2

# prepare minimal debian directory
cd cmd
mkdir debian

cat <<EOF > debian/control
Source: kubeadm
Section: misc
Priority: optional
Maintainer: Kubernetes Authors <kubernetes-dev+release@googlegroups.com>
Standards-Version: 3.9.8
Homepage: https://kubernetes.io
Package: kubeadm
Architecture: amd64
Depends: kubelet (>= $VERSION), kubectl (>= $VERSION)
Description: Kubernetes Cluster Bootstrapping Tool
 The Kubernetes command line tool for bootstrapping a Kubernetes cluster.
EOF

cat <<EOF > debian/rules
#!/usr/bin/make -f
%:
        dh \$@
EOF

chmod +x debian/rules

cat <<EOF > debian/changelog
kubeadm ($VERSION) unstable; urgency=medium
  * Test build.
 -- Kubernetes Authors <kubernetes-dev+release@googlegroups.com>  $(date -R)
EOF

cat <<EOF > debian/install
../_output/bin/kubeadm usr/bin
../build/debs/10-kubeadm.conf etc/systemd/system/kubelet.service.d/
EOF

echo 9 > debian/compat
cp ../build/debs/postinst debian/

# generate build script
cat <<EOF > ./build-kubeadm-deb.sh
apt-get update
apt-get install -y rsync dpkg-dev debhelper
cd /opt/k8s
make WHAT=cmd/kubeadm
cd cmd
dpkg-buildpackage -b
echo 'package content:'
dpkg -c ../kubeadm_${VERSION}_*.deb
EOF

cd ..

# run the build
docker run -it --rm -v$(pwd):/opt/k8s golang:1.11.5-stretch /bin/bash /opt/k8s/cmd/build-kubeadm-deb.sh

```

**Comment by bart0sh** [comment]
fpm produces more or less the same result:
```
$ make WHAT=cmd/kubeadm
$ mkdir -p pkg-build/{usr/bin,etc/systemd/system/kubelet.service.d/}
$ cp _output/bin/kubeadm pkg-build/usr/bin/
$ cp build/debs/10-kubeadm.conf pkg-build/etc/systemd/system/kubelet.service.d/

$ fpm -s dir -t deb -C ./pkg-build --name kubeadm --version 0.14.0~alpha.2 --post-install build/debs/postinst --maintainer 'Kubernetes Authors <kubernetes-dev+release@googlegroups.com>' --description "Kubernetes Cluster Bootstrapping Tool" --url https://kubernetes.io --vendor Kubernetes --license 'ASL 2.0'

$ dpkg -c kubeadm_0.14.0~alpha.2_amd64.deb 
drwxr-xr-x 0/0               0 2019-02-07 09:44 ./
drwxr-xr-x 0/0               0 2019-02-07 09:44 ./usr/
drwxr-xr-x 0/0               0 2019-02-07 09:44 ./usr/bin/
-rwxr-xr-x 0/0        36392288 2019-02-07 09:42 ./usr/bin/kubeadm
drwxr-xr-x 0/0               0 2019-02-07 09:44 ./usr/share/
drwxr-xr-x 0/0               0 2019-02-07 09:44 ./usr/share/doc/
drwxr-xr-x 0/0               0 2019-02-07 09:44 ./usr/share/doc/kubeadm/
-rw-r--r-- 0/0             173 2019-02-07 09:44 ./usr/share/doc/kubeadm/changelog.gz
drwxr-xr-x 0/0               0 2019-02-07 09:44 ./etc/
drwxr-xr-x 0/0               0 2019-02-07 09:44 ./etc/systemd/
drwxr-xr-x 0/0               0 2019-02-07 09:44 ./etc/systemd/system/
drwxr-xr-x 0/0               0 2019-02-07 09:44 ./etc/systemd/system/kubelet.service.d/
-rw-r--r-- 0/0             898 2019-02-07 09:43 ./etc/systemd/system/kubelet.service.d/10-kubeadm.conf

$ dpkg -I kubeadm_0.14.0~alpha.2_amd64.deb 
 new Debian package, version 2.0.
 size 11080718 bytes: control archive=922 bytes.
      54 bytes,     1 lines      conffiles            
     308 bytes,    11 lines      control              
     206 bytes,     3 lines      md5sums              
     693 bytes,    29 lines   *  postinst             #!/bin/sh
 Package: kubeadm
 Version: 0.14.0~alpha.2
 License: ASL 2.0
 Vendor: Kubernetes
 Architecture: amd64
 Maintainer: Kubernetes Authors <kubernetes-dev+release@googlegroups.com>
 Installed-Size: 35540
 Section: default
 Priority: extra
 Homepage: https://kubernetes.io
 Description: Kubernetes Cluster Bootstrapping Tool

$ ar p kubeadm_0.14.0~alpha.2_amd64.deb control.tar.gz | tar -zt
./
./conffiles
./control
./md5sums
./postinst
```

**Comment by neolit123** [comment]
@bart0sh could you please work with @Klaven (he is Marek Counts on slack) on figuring out if we really need the FPM tool?
on a quick look it seems like the tool reduces the amount of bash we need to write, which is always good.

**Comment by ncdc** [comment]
I've been building .rpms and .debs for Heptio Kubernetes Subscription for the past few months, and anything that can facilitate making package builds as easy as possible will get my vote. @Klaven & I are planning on syncing up on `fpm` today.

**Comment by hoegaarden** [comment]
I think one reason why @timothysc  wanted to go with a shared config (e.g. the config for nfpm) was, that we e.g. only bump versions, dependencies, in one place and it makes it through to all the packages (debs & rpms for now). When we use the native tooling for the package we need to change things in at least 2 places, we run into the issue where the package metadata diverges.
Sharing the config is not the only way to guard against such a skew.

Another thing to consider is creating packages for other platforms/architectures. I would guess that's not that big of a deal as long as you have means to compile the thing (which we have). I assume you can then just wrap those x-compiled binaries into packages for different platforms, no matter on which platform you do the packaging on?

**Comment by ncdc** [comment]
@hoegaarden 💯 for having a single location to manage package versions & dependencies

**Comment by ixdy** [comment]
awesome progress! two thoughts:
1. Building packages for other architecures gets a little trickier when you need to cross-compile cgo. We have crossbuild toolchains installed in the `kube-cross` image for this reason. (I think only `kubelet` is affected by this, though.)
2. Is there any way to automatically populate the version from git tags like `hack/lib/version.sh` does?

**Comment by ncdc** [comment]
@ixdy re 2, we definitely should be able to (that's what we do for HKS builds)

**Comment by bart0sh** [comment]
> When we use the native tooling for the package we need to change things in at least 2 places.

It depends on implementation. Here is an example script that builds deb&rpm using the same VERSION variable: https://gist.github.com/bartosh/3d11e2e4c71626e470f8df2b7592a741

Please, don't consider this as an argument against fpm or other tool. As I said above fpm produces the same result with much less effort, so why not to use it if it suits our requirements.

**Comment by ncdc** [comment]
@bart0sh that does still require that you specify the dependencies in 2 places, though 😄

**Comment by bart0sh** [comment]
@ncdc yes, it depends on implementation :)

@neolit123 
> on a quick look it seems like the tool reduces the amount of bash we need to write, which is always good.

It comes with a price of less flexibility though.

As for amount of shell scripting - it can be minimised by moving spec and Debian files into templates(either in their simplest form or full-featured like jinja or go template )

**Comment by hoegaarden** [comment]
> It comes with a price of less flexibility though.

I think we don't need much bash for creating a package with either `fpm` nor `nfpm`. I think we might need some tooling to build all the packages for all the platforms. That thing could be bash, but could also be golang.

I think we could go with e.g. the `nfpm` ("`nfpm`" and "`nfpm` config" here are just an examples, If you find other/better/nicer tools then `nfpm`, `fpm` also cool!) config file format and use `nfpm` for now. In case we find that `nfpm` is not good enough, we can always replace `nfpm` with native tooling and wrap that into a thin layer which reads the `nfpm` config and translates that to configuration for the native or any other tools -- but hidden from me, the caller of the package builder.

But then we need to maintain those layers. Which is fine if there are good reasons for that. But if we can postpone that until we really need to (and just use `nfpm` or other tools directly) I'd prefer that.

> As for amount of shell scripting - it can be minimised by moving spec and Debian files into templates(either in their simplest form or full-featured like jinja or go template )

I am not sure about that. I think I'd rather have plain specs and systemd files and whatnot, which do not require any templating anymore. I think I'd rather have a directory structure which holds specs and stuff and the `nfpm` config which defines overrides.

Something like this:
```
./kubesomething.nfpm.yaml
./99-kubesomething-systemd-unit.conf
./deb/arm64/99-kubesomething-systemd-unit.conf
./rpm/sparc64/99-kubesomething-systemd-unit.conf
```

Every package built would use the `./99-kubesomething-systemd-unit.conf` file, but in the `nfpm` config I can have overrides, which use a different one for the debian packages for arm64 and another different one for the rpm for sparc64.

**Comment by bart0sh** [comment]
@hoegaarden thanks for the explanations. sounds good to me.

BTW, How are we going to deal with build/debs and build/rpms content? Are we going to remove that in future or maintain nfpm configs in parallel?

Frankly, I'm not sure I fully understand the idea behind this issue. So far I only understood that we need to implement package building out of the source tree. However, the "big picture" (bazel builds, release builds, signing, etc) is still not very clear to me.

**Comment by hoegaarden** [comment]
@bart0sh My thinking is to remove all other package building mechanism. Not right away, but on the long run. When the new package building thing is finished and works (whatever that means) I think we want to deprecate all the other options and eventually remove them.

I tried to express that [in the KEP](https://github.com/kubernetes/enhancements/pull/707) but please note that the KEP is not merged / in implementable state. It is basically just my findings, ideas, ... with this issue as a starting point and with a lot of input from other people -- but it is nothing the community really agreed upon.

**Comment by bart0sh** [comment]
@hoegaarden interesting reading. thanks.

**Comment by hoegaarden** [comment]
[#73930 - bazel: initial support for cross-compilation](https://github.com/kubernetes/kubernetes/pull/73930) might also be interesting to you all.

**Comment by ibrasho** [comment]
Hey!

Gentle reminder that code freeze is this Thursday.

Since this is part of the 1.14 milestone, is it going to be resolved soon?

**Comment by hoegaarden** [comment]
/milestone 1.15
As discussed amongst @spiffxp @timothysc @hoegaarden and others

**Comment by BenTheElder** [comment]
/unassign
Also not going to have cycles for this. Doubtful other assignees will either @ixdy

**Comment by timothysc** [comment]
Update: 

On evaluation of fpm/nfpm we considered it a bit awkward and possibly more difficult to maintain then just building standard packages in k/k.  

Instead just move to clean it up and get bazel out of the generation business but just provide variable substitution where needed.

**Comment by josiahbjorgaard** [comment]
Since this issue is tagged for milestone v1.16, here is a gentle reminder that code freeze is starting in 9 days on August 29th. Is the issue still targeted for this release?

**Comment by xmudrii** [comment]
@timothysc What is the status of this issue? Should we move it to the next milestone?

**Comment by dims** [comment]
@xmudrii yes please

**Comment by xmudrii** [comment]
@dims Thanks for the update!

/milestone v1.17

**Comment by josiahbjorgaard** [comment]
Bug triage for 1.17 here. This issue has been open for a significant amount of time and since it is tagged for milestone 1.17, we want to let you know that the 1.17 code freeze is coming in less than one month on Nov. 18th. Will this issue be resolved before then?

**Comment by neolit123** [comment]
hopefully it can be resolved in future cycles.
/milestone v1.18

**Comment by zmlpjuran** [comment]
Hi guys, I hope that you have vendor specific rpm tags in the scope (ie: Vendor and Distribution).
Do you plan to fill it them statically or just in CI/CD?

**Comment by justaugustus** [comment]
This issue was originally opened at a time when the [`kubernetes/release` repo](https://github.com/kubernetes/release) was relatively unmaintained. I think it made a lot of sense then to have build tooling in k/k if it meant the likelihood of having clear owners and continued contributions to our overall all Release Engineering posture.

Since then, SIG Release has activated the @kubernetes/release-engineering subproject, which has made enormous improvements in the kubernetes/release content.

In https://github.com/kubernetes/kubernetes/pull/87585, we've removed the deb/rpm build definitions/specs from k/k.

In https://github.com/kubernetes/release/issues/1027, we track the required steps for creating a Go-based deb/rpm tool called `kubepkg`.

That said, I'm considering this issue closed.
If you have feature requests for `kubepkg`, please comment on the aforementioned issue.

/close
